Source/Task/timer_cpp.cpp

```diff
@@ -0,0 +1,187 @@
+    }
+};
+
+class TimerQueue
+{
+public:
+    bool LazyInit() noexcept;
+
+    void Add(PlatformTimer* timer, Deadline deadline) noexcept;
+    void Remove(PlatformTimer const* timer) noexcept;
+
+private:
+    void Worker() noexcept;
+
+    TimerEntry const& Peek() const noexcept;
+    TimerEntry Pop() noexcept;
+
+    std::mutex m_mutex;
+    std::condition_variable m_cv;
+    std::vector<TimerEntry> m_queue; // used as a heap
+    bool m_initialized = false;
+};
+
+namespace
+{
+
+std::once_flag g_timerQueueLazyInit;
+TimerQueue g_timerQueue;
+
+}
+
+bool TimerQueue::LazyInit() noexcept
+{
+    std::call_once(g_timerQueueLazyInit, [this]()
+    {
+        try
+        {
+            std::thread t([this]()
+            {
+                Worker();
+            });
+            t.detach();
+
+            m_initialized = true;
+        }
+        catch (...)
+        {
+            m_initialized = false;
+        }
+    });
+
+    return m_initialized;
+}
+
+void TimerQueue::Add(PlatformTimer* timer, Deadline deadline) noexcept
+{
+    {
+        std::lock_guard<std::mutex> lock{ m_mutex };
+
+        m_queue.emplace_back(deadline, timer);
+        std::push_heap(m_queue.begin(), m_queue.end(), TimerEntryComparator{});
+    }
+    m_cv.notify_all();
+}
+
+void TimerQueue::Remove(PlatformTimer const* timer) noexcept
+{
+    std::lock_guard<std::mutex> lock{ m_mutex };
+
+    for (auto& entry : m_queue)
+    {
+        if (entry.Timer == timer)
+        {
+            entry.Timer = nullptr;
+        }
+    }
+}
+
+void TimerQueue::Worker() noexcept
+{
+    std::unique_lock<std::mutex> lock{ m_mutex };
+    while (true)
+    {
+        while (!m_queue.empty())
+        {
+            Deadline next = Peek().When;
+            if (std::chrono::high_resolution_clock::now() > next)
+            {
+                break;
+            }
+
+            TimerEntry entry = Pop();
+
+            // release the lock while invoking the callback, just in case timer
+            // gets destroyed on this thread or readds itself in the callback
+            lock.unlock();
+            if (entry.Timer) // Timer is set to nullptr if the entry is removed
+            {
+                entry.Timer->OnDeadline();
+            }
+            lock.lock();
+        }
+
+        if (!m_queue.empty())
+        {
+            Deadline next = Peek().When;
+            m_cv.wait_until(lock, next);
+        }
+        else
+        {
+            m_cv.wait(lock);
+        }
+    }
+}
+
+TimerEntry const& TimerQueue::Peek() const noexcept
+{
+    return m_queue.front();
+}
+
+TimerEntry TimerQueue::Pop() noexcept
+{
+    TimerEntry e = m_queue.front();
+    std::pop_heap(m_queue.begin(), m_queue.end(), TimerEntryComparator{});
+    m_queue.pop_back();
+    return e;
+}
+
+PlatformTimer::PlatformTimer(void* context, PlatformTimerCallback* callback) noexcept :
+    m_context{ context },
+    m_callback{ callback },
+    m_valid{ g_timerQueue.LazyInit() }
```

Comment:
This burns a thread on Android always, even if you never use timers.  Should lazy init be called in Start and make Start return a bool for success?  Or maybe provide a general Init method on PlatformTimer that is called when we first need it.  I think it will be rare for folks to use these.

Reply:
Good points. I'll see what I can do to make this better

---
In reply to: [192512998](https://github.com/Microsoft/libHttpClient/pull/181#discussion_r192512998) [](ancestors = 192512998)

Reply:
Will fix in a second PR

---
In reply to: [192515854](https://github.com/Microsoft/libHttpClient/pull/181#discussion_r192515854) [](ancestors = 192515854,192512998)